### PR TITLE
Upgrade Graal to 21.0

### LINKED ISSUE #1296: Bump GraalVM Version to 20.3
### Summary
Oracle has just released the [new version](https://github.com/graalvm/graalvm-ce-builds/releases/tag/vm-20.3.0) of GraalVM, and we need to update our project and build to use it.

### Value
We won't get stuck on an older and unsupported version of GraalVM.

### Specification

- [ ] Update the build to use 20.3.
- [ ] Fix all deprecations by updating the interpreter to use the replacements where necessary.
- [ ] Check if the MethodHandle support changes in the Native Image are enough to remove the substitutions we have.
  - Remove `ReplacementStatics` and `ReplacementAkkaSpinWait` and verify if the native images still build and pass the tests.

### Acceptance Criteria & Test Cases

- [ ] The build uses the new version of Graal and there aren't any deprecation warnings.

### LINKED ISSUE #1481: Implement `to_display` for Errors.
### Summary

The dataflow errors lack a general way to convert them to an user-friendly message. The `Error` type should implement `to_display` method.

### Value

IDE could easily display an error message inside visualization.

### PR DESCRIPTION
**Pull request by iamrecursion**
### Pull Request Description
We use the hotfix version `21.0.0.2`.

Closes #1481.
Closes #1296.

### Important Notes

- Fixes exception handling with polyglot stuff.
- Fixes some CI issues.
- Added notes on the `add-exports` stuff.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java [hunk 1]
```diff
@@ -95,10 +101,13 @@ public Builtins(Context context) {
     nothing = new AtomConstructor("Nothing", scope).initializeFields();
     number = new Number(language, scope);
     ordering = new Ordering(language, scope);
+    panic = new AtomConstructor("Panic", scope).initializeFields();
     polyglot = new Polyglot(language, scope);
     resource = new Resource(language, scope);
     system = new System(language, scope);
     text = new Text(language, scope);
+    thunk = new AtomConstructor("Thunk", scope).initializeFields();
```
Comment (kustosz) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java]: This should not exist

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java [hunk 2]
```diff
@@ -301,4 +309,48 @@ public ModuleScope getScope() {
...
+        return nothing.newInstance();
+      case Constants.NUMBER:
+        return number.getNumber().newInstance();
+      case Constants.PANIC:
+        return panic.newInstance();
+      case Constants.REF:
+        return mutable.ref().newInstance();
+      case Constants.TEXT:
+        return text.getText().newInstance();
+      case Constants.THUNK:
```
Comment (kustosz) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java]: This should not exist. It is incorrect to expect a thunk, ever. The runtime may decide to pass an evaluated value in the place of a thunk, without notice.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java [hunk 3]
```diff
@@ -95,10 +101,13 @@ public Builtins(Context context) {
     nothing = new AtomConstructor("Nothing", scope).initializeFields();
     number = new Number(language, scope);
     ordering = new Ordering(language, scope);
+    panic = new AtomConstructor("Panic", scope).initializeFields();
     polyglot = new Polyglot(language, scope);
     resource = new Resource(language, scope);
     system = new System(language, scope);
     text = new Text(language, scope);
+    thunk = new AtomConstructor("Thunk", scope).initializeFields();
+    unresolvedSymbol = new AtomConstructor("Unresolved_Symbol", scope).initializeFields();
```
Comment (kustosz) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java]: This should also not exist (yet :P)

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java [hunk 4]
```diff
@@ -95,10 +101,13 @@ public Builtins(Context context) {
     nothing = new AtomConstructor("Nothing", scope).initializeFields();
     number = new Number(language, scope);
     ordering = new Ordering(language, scope);
+    panic = new AtomConstructor("Panic", scope).initializeFields();
```
Comment (kustosz) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java:102]: This already exists.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java [hunk 5]
```diff
@@ -301,4 +309,48 @@ public ModuleScope getScope() {
...
+        return number.getNumber().newInstance();
+      case Constants.PANIC:
+        return panic.newInstance();
+      case Constants.REF:
+        return mutable.ref().newInstance();
+      case Constants.TEXT:
+        return text.getText().newInstance();
+      case Constants.THUNK:
+        return thunk.newInstance();
+      case Constants.UNRESOLVED_SYMBOL:
```
Comment (kustosz) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java]: Again, this should not be assertable. A user should never expect an unresolved symbol, just a function.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/error/PanicSentinel.java [hunk 6]
```diff
@@ -12,10 +16,10 @@
  * <p>This tracing is enabled by the active intervention of the runtime instrumentation, and does
  * not function in textual mode.
  */
+@ExportLibrary(value = InteropLibrary.class, delegateTo = "panic")
 @ExportLibrary(MethodDispatchLibrary.class)
-public class PanicSentinel extends RuntimeException implements TruffleException {
-  private final PanicException panic;
-  private final Node location;
+public class PanicSentinel extends AbstractTruffleException {
```
Comment (kustosz) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/error/PanicSentinel.java:20]: This should not be an `AbstractTruffleException` nor export the exceptions library. Notice how this makes the panic sentinels handlable in language, which is exactly opposite to what we want.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/TopLevelScope.java [hunk 7]
```diff
@@ -22,12 +22,16 @@
 import java.util.*;
 import java.util.stream.Collectors;
 
+// TODO [AA] 21.0.0 migration
```
Comment (kustosz) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/TopLevelScope.java]: .

**Review by kustosz — CHANGES_REQUESTED** [review]

**Review by kustosz — APPROVED** [review]

**Review by radeusgd — APPROVED** [review]
Looks good to me, I just have two minor questions.
Accepting conditionally on the release workflow succeeding on staging.

Also I think we need to remember to create an issue to improve the CI timing in the future, as there are places in the current workarounds that may be improved (only cleaning the instruments, avoiding additional rebuilds by only using the `build*Distribution` commands instead of `assembly` etc.)

File: .github/workflows/scala.yml [hunk 8]
```diff
@@ -8,30 +8,19 @@ on:
...
-    .github/PULL_REQUEST_TEMPLATE.md
-    .github/CODEOWNERS
-    .github/ISSUE_TEMPLATE/*
-    /doc/*
-    .gitignore
-    .scalafmt.conf
-    CODE_OF_CONDUCT.md
-    CONTRIBUTING.md
-    LICENSE
-    README.md
```
[the remark this answers is not in the document]
Comment (iamrecursion) [reply .github/workflows/scala.yml:28]: It is. It's unused and I'm not sure why it's been around for so long!

**Pull request merged**